### PR TITLE
Integration with Spring Cloud Sleuth

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
### CorrelationIdFilter

There is no more `CorrelationIdFilter`. In tests instead of sth like

```
@ContextConfiguration(classes = [Application], loader = SpringApplicationContextLoader)
class MicroserviceMvcWiremockSpec extends MvcWiremockIntegrationSpec {

    @Override
    protected void configureMockMvcBuilder(ConfigurableMockMvcBuilder mockMvcBuilder) {
        super.configureMockMvcBuilder(mockMvcBuilder)
        mockMvcBuilder.addFilter(new CorrelationIdFilter())
    }
}

```

you have to have

```
import com.ofg.infrastructure.base.MvcWiremockIntegrationSpec
import org.springframework.cloud.sleuth.instrument.web.TraceFilter
import com.ofg.infrastructure.web.correlationid.HeadersSettingFilter
import com.ofg.twitter.Application
import org.springframework.beans.factory.annotation.Autowire
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.actuate.trace.Trace
import org.springframework.boot.test.SpringApplicationContextLoader
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.web.servlet.setup.ConfigurableMockMvcBuilder

@ContextConfiguration(classes = [Application], loader = SpringApplicationContextLoader)
class MicroserviceMvcWiremockSpec extends MvcWiremockIntegrationSpec {

    @Autowired Trace trace
    
    @Override
    protected void configureMockMvcBuilder(ConfigurableMockMvcBuilder mockMvcBuilder) {
        super.configureMockMvcBuilder(mockMvcBuilder)
        mockMvcBuilder.addFilter(new HeadersSettingFilter(), new TraceFilter(trace))
    }
}

```

### Headers are not set for "first" request

Had to add `com.ofg.infrastructure.web.correlationid.HeadersSettingFilter` that creates headers for `spanId` and `traceId` if there are not set. Without those `TraceFilter` is not setting `traceId and `spanId` to response

### TraceCallable and TraceRunnable should clean after itself

Kudos to @nurkiewicz for help on the analysis. Both classes are working on a separate thread pool which is shared by a number of threads. After either of them finishes it's job it should remove the Span from current ThreadLocal. The same with `TraceCommand`.

### Exception should not be thrown when there is a tracing issue

For now we're wrapping it in `ExceptionCatchingTraceScope` that wraps `TraceScope` in `com.ofg.infrastructure.tracing.TracingConfiguration.ExceptionCatchingTraceScope`

### Prefix not appended

With Spring Boot 1.3.0.RELEASE Spring Cloud Zookeeper fails to combine the `spring.cloud.zookeeper.prefix` with `spring.application.name`

## Issues

Swagger is not compatible with Spring Boot 1.3.0.RELEASE  and Spring 4.2.3.RELEASE - https://github.com/springfox/springfox/issues/1055

### Workaround

I've repackaged Swagger and changed this class as follows:

```
package repackaged.com.mangofactory.swagger.readers.operation;

import com.mangofactory.swagger.scanners.RequestMappingContext;

/**
 * Terrible hack until they fix https://github.com/springfox/springfox/issues/1055
 */
public class OperationDeprecatedReader implements RequestMappingReader {
  @Override
  public void execute(RequestMappingContext context) {
    context.put("deprecated", String.valueOf(false));
  }
}
```

### CODE REVIEW

**Comment by marcingrzejszczak** [comment]
Done in https://github.com/4finance/micro-infra-spring/commit/dcd7cd81186b710982a77becea5add3c855c9c1f

File: .travis.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ sudo: false
 
 before_install:
   - "echo $JAVA_OPTS"
-  - "export JAVA_OPTS='-Xmx1024m -XX:MaxPermSize=384m'"
+  - "export JAVA_OPTS='-Xmx2048m -XX:MaxPermSize=500m'"
```
Comment (szpak) [inline .travis.yml]: Is it really needed? Maybe change in `jvmArgs` in tests is enough?

File: gradle/runBootMicroserviceAcceptanceTests.sh [hunk 2]
```diff
@@ -16,10 +16,11 @@ dumpCurrentMicroInfraSpringVersionToFile
 
 git clone https://github.com/4finance/boot-microservice.git
 cd boot-microservice
+git checkout tech/integration_with_sleuth
 updateMicroInfraSpringVersionInConfigurationFile
 ./gradlew clean check uptodate --stacktrace --continue --info --no-daemon
 
 git reset --hard
-git checkout boot-microservice-gui
+git checkout tech/integration_with_sleuth_gui
```
Comment (szpak) [inline gradle/runBootMicroserviceAcceptanceTests.sh]: Please remember to restore the original branch when merged in boot-microservice.

File: micro-deps/micro-deps-spring-config/src/test/groovy/com/ofg/infrastructure/discovery/ApplicationContextStartupSpec.groovy [hunk 3]
```diff
@@ -2,18 +2,27 @@ package com.ofg.infrastructure.discovery
...
     def 'should fail to start application context if resource is missing when default bean is missing deps'() {
         given:
-            TestingServer testingServer = new TestingServer(2181)
+            TestingServer testingServer
+            Integer port
+            new AvailablePortScanner(9000, 9500).tryToExecuteWithFreePort { int freePort ->
+                testingServer = new TestingServer(freePort)
+                port = freePort
+            }
+            System.setProperty('service.resolver.url', "localhost:$port")
```
Comment (szpak) [inline micro-deps/micro-deps-spring-config/src/test/groovy/com/ofg/infrastructure/discovery/ApplicationContextStartupSpec.groovy]: `port` could be returned from the closure

File: .travis.yml (hunk 1 above)
Comment (marcingrzejszczak) [inline .travis.yml]: Fixed

File: gradle/runBootMicroserviceAcceptanceTests.sh (hunk 2 above)
Comment (marcingrzejszczak) [inline gradle/runBootMicroserviceAcceptanceTests.sh]: Brought back

File: micro-deps/micro-deps-spring-config/src/test/groovy/com/ofg/infrastructure/discovery/ApplicationContextStartupSpec.groovy (hunk 3 above)
Comment (marcingrzejszczak) [inline micro-deps/micro-deps-spring-config/src/test/groovy/com/ofg/infrastructure/discovery/ApplicationContextStartupSpec.groovy]: Done in https://github.com/4finance/micro-infra-spring/commit/41841e01cdf533387f702c81aafdfa74535a2a6e

**Pull request closed without merging**